### PR TITLE
 ISPN-11280 ConcurrentModificationException in ConditionFuture

### PR DESCRIPTION
**Pull request by danberindei**
https://issues.redhat.com/browse/ISPN-11280
https://issues.redhat.com/browse/ISPN-11296

### CODE REVIEW

File: core/src/main/java/org/infinispan/util/concurrent/ConditionFuture.java [hunk 1]
```diff
@@ -96,24 +99,31 @@ public void updateAsync(T value, Executor executor) {
 
       try {
          executor.execute(() -> checkConditions(value));
-      } catch (Exception e) {
+      } catch (Throwable t) {
          for (Data data : futures.values()) {
```
Comment (pruivo) [inline core/src/main/java/org/infinispan/util/concurrent/ConditionFuture.java]: doesn't need to be synchronized as well? `futures.values()`

Comment (pruivo) [reply core/src/main/java/org/infinispan/util/concurrent/ConditionFuture.java]: ps. same for `stop()`

Comment (danberindei) [reply core/src/main/java/org/infinispan/util/concurrent/ConditionFuture.java]: Done, good eyes Pedro!

**Comment by pruivo** [comment]
integrated! thanks @danberindei !

File: core/src/main/java/org/infinispan/util/concurrent/ConditionFuture.java [hunk 2]
```diff
@@ -96,35 +99,58 @@ public void updateAsync(T value, Executor executor) {
...
+      for (Data data : completed) {
+         data.complete(null);
+      }
    }
 
    public void stop() {
       running = false;
       lastValue = null;
+
+      List<CompletableFuture<?>> completed;
```
Comment (johnou) [inline core/src/main/java/org/infinispan/util/concurrent/ConditionFuture.java]: @danberindei how about something like this to cut down on the sync scope?

```java
        final List<Data> copy;
        synchronized (futures) {
            copy = new ArrayList<>(futures.values());
            futures.clear();
        }
        for (Data data : copy) {
            data.cancelFuture.cancel(false);
        }
```

File: core/src/main/java/org/infinispan/util/concurrent/ConditionFuture.java [hunk 3]
```diff
@@ -96,35 +99,58 @@ public void updateAsync(T value, Executor executor) {
 
       try {
          executor.execute(() -> checkConditions(value));
-      } catch (Exception e) {
-         for (Data data : futures.values()) {
-            data.cancelFuture.cancel(false);
-            data.completeExceptionally(e);
+      } catch (Throwable t) {
+         List<CompletableFuture<?>> completed;
+         synchronized (futures) {
```
Comment (johnou) [inline core/src/main/java/org/infinispan/util/concurrent/ConditionFuture.java]: same sync optimisation can be applied here too.

**Pull request merged**